### PR TITLE
Add expire time for nodeInfo cache items

### PR DESCRIPTION
**Pull request by yaroslava-serdiuk**
#### Which component this PR applies to?

cluster-autoscaler

#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
Adding expire time for nodeInfo cache items. Expire time could be set via flag `--node-info-cache-expire-time`, by default there is no expire time (old behavior).

Currently, item in the node info cache is updated only if there are any nodes in the node group. If node group is scaled down to 0, node info cache will store last node and use it in scale up algorithm. Thus if the last node is a wrong representative of the node group (for example it has custom taints), then Cluster Autoscaler may struggle to scale up this node group.

```release-note
Added `--node-info-cache-expire-time` flag for Cluster Autoscaler to specify the time to store node info cache items.
```

### CODE REVIEW

File: cluster-autoscaler/processors/nodeinfosprovider/mixed_nodeinfos_processor.go [hunk 1]
```diff
@@ -36,17 +36,31 @@ import (
...
+func NewMixedTemplateNodeInfoProvider(ttl *time.Duration) *MixedTemplateNodeInfoProvider {
 	return &MixedTemplateNodeInfoProvider{
-		nodeInfoCache: make(map[string]*schedulerframework.NodeInfo),
+		nodeInfoCache: make(map[string]cacheItem),
+		ttl:           ttl,
+	}
+}
+
+func (p *MixedTemplateNodeInfoProvider) isCacheItemExpired(added time.Time) bool {
+	if p.ttl == nil || *p.ttl < 0 {
```
Comment (x13n) [inline cluster-autoscaler/processors/nodeinfosprovider/mixed_nodeinfos_processor.go:60]: Why do we need both nil duration and negative duration to achieve the same thing as setting the TTL to an arbitrarily large value?

File: cluster-autoscaler/processors/processors.go [hunk 2]
```diff
@@ -77,7 +77,6 @@ func DefaultProcessors() *AutoscalingProcessors {
 		NodeInfoProcessor:          nodeinfos.NewDefaultNodeInfoProcessor(),
 		NodeGroupConfigProcessor:   nodegroupconfig.NewDefaultNodeGroupConfigProcessor(),
 		CustomResourcesProcessor:   customresources.NewDefaultCustomResourcesProcessor(),
-		TemplateNodeInfoProvider:   nodeinfosprovider.NewDefaultTemplateNodeInfoProvider(),
```
Comment (x13n) [inline cluster-autoscaler/processors/processors.go:80]: Why did you remove this line?

File: cluster-autoscaler/processors/nodeinfosprovider/mixed_nodeinfos_processor.go [hunk 3]
```diff
@@ -36,17 +36,31 @@ import (
 
 const stabilizationDelay = 1 * time.Minute
 
+type cacheItem struct {
```
Comment (x13n) [inline cluster-autoscaler/processors/nodeinfosprovider/mixed_nodeinfos_processor.go:39]: nit: You could embed `*schedulerframework.NodeInfo` here.

File: cluster-autoscaler/main.go [hunk 4]
```diff
@@ -185,6 +186,7 @@ var (
 
 	emitPerNodeGroupMetrics  = flag.Bool("emit-per-nodegroup-metrics", false, "If true, emit per node group metrics.")
 	debuggingSnapshotEnabled = flag.Bool("debugging-snapshot-enabled", false, "Whether the debugging snapshot of cluster autoscaler feature is enabled")
+	nodeInfoCacheExpireTime  = flag.Duration("node-info-cache-expire-time", -1*time.Minute, "Node Info cache expire time for each item. If the value isn't set or negative, the cache items won't be expired")
```
Comment (x13n) [inline cluster-autoscaler/main.go:189]: nit: You could default to something like 100 years to avoid special casing negative values.

**Review by x13n — CHANGES_REQUESTED** [review]

File: cluster-autoscaler/processors/processors.go (hunk 2 above)
Comment (yaroslava-serdiuk) [reply cluster-autoscaler/processors/processors.go:80]: I don't think it is default processors after introducing an ttl time parameter. In order to not pass a ttl time to DefaultProcessors() function I initialise TemplateNodeInfoProvider separately in main.go

File: cluster-autoscaler/processors/nodeinfosprovider/mixed_nodeinfos_processor.go [hunk 5]
```diff
@@ -36,17 +36,31 @@ import (
...
+func NewMixedTemplateNodeInfoProvider(ttl *time.Duration) *MixedTemplateNodeInfoProvider {
 	return &MixedTemplateNodeInfoProvider{
-		nodeInfoCache: make(map[string]*schedulerframework.NodeInfo),
+		nodeInfoCache: make(map[string]cacheItem),
+		ttl:           ttl,
+	}
+}
+
+func (p *MixedTemplateNodeInfoProvider) isCacheItemExpired(added time.Time) bool {
+	if p.ttl == nil {
```
Comment (x13n) [inline cluster-autoscaler/processors/nodeinfosprovider/mixed_nodeinfos_processor.go:60]: nit: Why do you use a pointer at all? If `ttl` was a regular value you wouldn't have to do this.

Comment (yaroslava-serdiuk) [reply cluster-autoscaler/processors/nodeinfosprovider/mixed_nodeinfos_processor.go]: I left pointer for test that mocks MixedTemplateNodeInfoProvider, so there is no need to pass a specific parameter there. However it should be handled during provider creation.

File: cluster-autoscaler/processors/nodeinfosprovider/mixed_nodeinfos_processor_test.go [hunk 6]
```diff
@@ -187,10 +191,10 @@ func TestGetNodeInfosForGroupsCache(t *testing.T) {
 	// Check cache
 	cachedInfo, found := niProcessor.nodeInfoCache["ng1"]
 	assert.True(t, found)
-	assertEqualNodeCapacities(t, ready1, cachedInfo.Node())
+	assertEqualNodeCapacities(t, ready1, cachedInfo.NodeInfo.Node())
```
[the remark this answers is not in the document]
Comment (yaroslava-serdiuk) [reply cluster-autoscaler/processors/nodeinfosprovider/mixed_nodeinfos_processor_test.go]: Yes, sorry missed few places.

File: cluster-autoscaler/processors/processors.go (hunk 2 above)
Comment (x13n) [reply cluster-autoscaler/processors/processors.go:80]: I understand, but now you've changed the default implementation so that it always has a nil `TemplateNodeInfoProvider`, so anyone who used to rely on this will suddenly get a nil pointer (and most likely `panic` down the road). I suggest either keeping it initialized with some default TTL value (like 10 years) or mention breaking change in a release note.

File: cluster-autoscaler/processors/nodeinfosprovider/mixed_nodeinfos_processor.go (hunk 5 above)
Comment (x13n) [reply cluster-autoscaler/processors/nodeinfosprovider/mixed_nodeinfos_processor.go]: So you have different state for tests and a different one for actual execution. If you used a const like `veryLongTime = 123456 * time.Hour` in the tests, you wouldn't need special casing `nil`.

Comment (yaroslava-serdiuk) [reply cluster-autoscaler/processors/nodeinfosprovider/mixed_nodeinfos_processor.go]: I wouldn't say that state for test is different, a nil pointer correspond to default value — the behavior is identical as if --node-info-cache-expire-time flag is not set. https://github.com/kubernetes/autoscaler/pull/4669/files#diff-ad9d6ea8aa127e0dc3b271f9b5761df7ba18f9fd63ffd0a7f69a6886c86ef317R54
I think that if new test is introduced that doesn't test nodeInforCache directly but use it, then the writer shouldn't come up with a new veryLongTime variable. 
Please, explain if my line of thought is wrong. 
(Also, after your argument I used nil pointer for initialising in DefaultProcessor(), which is used not only for test.)

File: cluster-autoscaler/processors/processors.go (hunk 2 above)
Comment (yaroslava-serdiuk) [reply cluster-autoscaler/processors/processors.go:80]: Got it, thanks. I returned TemplateNodeInfoProvider.
